### PR TITLE
refactored d2input, abstracted ebiten to backend interface

### PR DESCRIPTION
**Pull request by dknuth0101**
main.go is currently hardcoded to init input system with ebiten as backend

### CODE REVIEW

File: d2core/d2input/d2input_ebiten/backend_ebiten.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"github.com/OpenDiablo2/OpenDiablo2/d2core/d2input/keyboard"
+	"github.com/OpenDiablo2/OpenDiablo2/d2core/d2input/mouse"
+
+	"github.com/hajimehoshi/ebiten"
+	"github.com/hajimehoshi/ebiten/inpututil"
+)
+
+func keyToEbiten(k keyboard.Key) ebiten.Key {
+	return ebiten.Key(k)
```
Comment (FooSoft) [inline d2core/d2input/d2input_ebiten/backend_ebiten.go]: Does this just not assume that all of the constants are the same between ebiten and the values that you added? How will this work when you try to add another input backend?

File: d2core/d2input/keyboard/keyboard.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+	KeySlash
+	KeySpace
+	KeyTab
+	KeyUp
+	KeyAlt
+	KeyControl
+	KeyShift
+)
+
+var Keys = [...]Key{
```
Comment (FooSoft) [inline d2core/d2input/keyboard/keyboard.go]: Don't use variables for things that are actually supposed to be constants.

Comment (dknuth0101) [reply d2core/d2input/keyboard/keyboard.go]: I don't thing the array can be const. It's just there so we can range over the keys elsewhere. How should it be done as a const?

Comment (dknuth0101) [reply d2core/d2input/keyboard/keyboard.go]: Okay, I've addressed this in my recent commits.

File: d2core/d2input/d2input_ebiten/backend_ebiten.go (hunk 1 above)
Comment (dknuth0101) [reply d2core/d2input/d2input_ebiten/backend_ebiten.go]: A map is now explicitly defined, and not exported. It's now part of init for the input backend.

**Comment by dknuth0101** [comment]
hmm... the midi driver breaks the build, the build environment needs to have libs for portmidi. I looked into using other midi drivers, but same issue.

**Pull request closed without merging**